docs/source/daml/intro/3_Data.rst

```diff
@@ -0,0 +1,206 @@
+
+Throughout the scenario, ``eq_record`` never changes. The expression ``"Zero" :: eq_record.my_list`` doesn't change the list in-place, but creates a new list, which is ``eq_record.my_list`` with an extra element in the beginning.
+
+.. _contract_keys:
+
+Contract keys
+-------------
+
+DAML's type system lets you store richly structured data on DAML templates, but just like most database schemas have more than one table, DAML contract models often have multiple templates that reference each other. For example, you may not want to store your bank and account information on each individual cash balance, but instead store those separately.
+
+You have already met the type ``ContractId a``, which references a contract of type ``a``. The below shows a contract model where ``Account`` is split out into a separate template and referenced by ``ContractId``, but it also highlights a big problem with that kind of reference: Just like data, contracts are immutable. They can only be created and archived. That makes contract ids very unstable, and can cause stale references.
+
+.. literalinclude:: daml/3_Data/IDRef.daml
+  :language: daml
+  :start-after: -- ID_REF_TEST_BEGIN
+  :end-before: -- ID_REF_TEST_END
+
+Note the use of the ``fetch`` function in the above, which retrieves the arguments of an active contract from its contract id. Also take note that for the first time, the party submitting a transaction is doing more than one thing as part of that transaction. To create ``new_account``, the accountant fetches the arguments of the old account, archives the old account and creates a new account, all in one transaction. More on building transactions in :doc:`7_Composing`.
+
+DAML allows the definition of stable keys using the ``key`` and ``maintainer`` keywords. ``key`` defines the primary key of a template, with a uniqueness constraint on that key and the ability to look up contracts by key.
+
+.. literalinclude:: daml/3_Data/Keys.daml
+  :language: daml
+  :start-after: -- KEY_TEST_BEGIN
+  :end-before: -- KEY_TEST_END
+
+Since DAML is a distributed system, there is no global entity that can guarantee uniqueness, which is why each ``key`` expression must come with a ``maintainer`` expression. ``maintainer`` takes one or several parties, all of which have to be signatories of the contract and be part of the key. That way the index can be partitioned amongst sets of maintainers, and each set of maintainers can independently ensure the uniqueness constraint on their piece of the index.
```

Comment:
```suggestion
Since DAML is designed to run on distributed systems, you have to assume that there is no global entity that can guarantee uniqueness, which is why each ``key`` expression must come with a ``maintainer`` expression. ``maintainer`` takes one or several parties, all of which have to be signatories of the contract and be part of the key. That way the index can be partitioned amongst sets of maintainers, and each set of maintainers can independently ensure the uniqueness constraint on their piece of the index.
```